### PR TITLE
[CIFuzz] Affected fuzzers

### PR DESCRIPTION
**Pull request by Leo-Neat**
Add feature to only run fuzzers that are affected by the code change. This change prevents valuable time being wasted on fuzzing unchanged code.

### CODE REVIEW

File: infra/cifuzz/cifuzz.py [hunk 1]
```diff
@@ -311,7 +312,8 @@ def get_files_covered_by_target(latest_cov_info, target_name, src_in_docker):
     ValueError: When the src_in_docker is not defined.
   """
   if not src_in_docker:
-    raise ValueError('Project base is not defined. Can\'t get coverage')
+    logging.error('Project base is not defined. Can\'t get coverage')
```
Comment (oliverchang) [inline infra/cifuzz/cifuzz.py]: what's the reason for this change?

File: infra/cifuzz/cifuzz.py [hunk 2]
```diff
@@ -326,9 +328,14 @@ def get_files_covered_by_target(latest_cov_info, target_name, src_in_docker):
 
   affected_file_list = []
   for file in coverage_per_file:
-    if file['filename'].startswith(
-        src_in_docker) and file['summary']['regions']['count']:
-      affected_file_list.append(file['filename'].replace(src_in_docker, ''))
+    if not file['filename'].startswith(src_in_docker):
```
Comment (oliverchang) [inline infra/cifuzz/cifuzz.py]: Since we're doing path comparisons here, we should do os.path.normpath on both for robustness, in case there are redundant /../ etcs in either one.

File: infra/cifuzz/cifuzz.py [hunk 3]
```diff
@@ -320,24 +323,70 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
 
-    relative_path = file['filename'].replace(oss_fuzz_project_base, '')
+    relative_path = file['filename'].replace(src_in_docker, '')
     affected_file_list.append(relative_path)
   if not affected_file_list:
     return None
   return affected_file_list
 
 
+def keep_affected_fuzzers(project_name, out_dir, files_changed, src_in_docker):
```
Comment (oliverchang) [inline infra/cifuzz/cifuzz.py]: nit: probably better named as "remove_unaffected_fuzzers" to more explicitly indicate what this function is doing (removing).

File: infra/cifuzz/cifuzz.py [hunk 4]
```diff
@@ -320,24 +323,70 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
+  if not affected_fuzzers:
+    logging.info('Not using affected fuzzers.')
+    return
+  logging.info(
+      'Using affected fuzzers.\n The following fuzzers are affected: %s',
+      ' '.join(affected_fuzzers))
+
+  # Remove all the fuzzers that are not affected
+  for file in os.listdir(out_dir):
+    if file not in affected_fuzzers:
```
Comment (oliverchang) [inline infra/cifuzz/cifuzz.py]: nit: prefer early exit to avoid additional indentation.

i.e.

```python
if file in affected_fuzzers:
  continue

try:
  ...
```

File: infra/cifuzz/cifuzz.py [hunk 5]
```diff
@@ -320,24 +323,70 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
+  for fuzzer in fuzzer_paths:
+    covered_files = get_files_covered_by_target(latest_cov_report_info,
+                                                os.path.basename(fuzzer),
+                                                src_in_docker)
+
+    if covered_files and set(covered_files) & set(files_changed):
+      affected_fuzzers.append(os.path.basename(fuzzer))
+
+  if not affected_fuzzers:
+    logging.info('Not using affected fuzzers.')
```
Comment (oliverchang) [inline infra/cifuzz/cifuzz.py]: This message seems a bit confusing to me. Is this because not having affected fuzzers indicates there is something wrong with coverage and we should bail?  If so, let's make the message more clear e.g. 'No affected fuzzers detected, keeping all as fallback'. or something similar.

File: infra/cifuzz/cifuzz.py (hunk 1 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: On all other errors/issues I have been returning None so I wanted to maintain consistency. I also would just end up catching it in the keep_affected_fuzzers method and then doing the same thing as if it returned None.

File: infra/cifuzz/cifuzz.py (hunk 3 above)
Comment (jonathanmetzman) [reply infra/cifuzz/cifuzz.py]: +1

File: infra/cifuzz/cifuzz.py [hunk 6]
```diff
@@ -163,6 +163,10 @@ def build_fuzzers(project_name,
   if helper.docker_run(command):
     logging.error('Building fuzzers failed.')
     return False
+
+  # Removes non-affected fuzzers.
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz.py]: Also, could you drop the s in "Removes"

File: infra/cifuzz/cifuzz.py [hunk 7]
```diff
@@ -320,24 +323,70 @@ def get_files_covered_by_target(latest_cov_info, target_name,
     return None
 
   # Cases like curl there is /src/curl and /src/curl_fuzzers/ are handled.
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz.py]: Could you change this comment to some thing like "Make sure cases like /src/curl and /src/curl/ are both handled."

File: infra/cifuzz/cifuzz.py (hunk 2 above)
Comment (jonathanmetzman) [reply infra/cifuzz/cifuzz.py]: +1.

File: infra/cifuzz/cifuzz.py [hunk 8]
```diff
@@ -320,24 +323,70 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
+  """Keeps only the affected fuzzers in the out directory.
+
+  Args:
+    project_name: The name of the relevant OSS-Fuzz project.
+    out_dir: The location of the fuzzer binaries.
+    files_changed: A list of files changed from origin.
+    src_in_docker: The location of the source dir in the docker image.
+  """
+  if not files_changed:
+    logging.info('No files changed from origin.')
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz.py]: Under what circumstances could this happen? Would it be a bug if this happened that we should log as error?

File: infra/cifuzz/cifuzz.py [hunk 9]
```diff
@@ -320,24 +323,70 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
+  if not latest_cov_report_info:
+    logging.error('Could not download latest coverage report.')
+    return
+  affected_fuzzers = []
+  for fuzzer in fuzzer_paths:
+    covered_files = get_files_covered_by_target(latest_cov_report_info,
+                                                os.path.basename(fuzzer),
+                                                src_in_docker)
+
+    if covered_files and set(covered_files) & set(files_changed):
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz.py]: I didn't know the ampersand got the intersection of two sets in Python.
Can you do `set(covered_files).intersection(set(files_changed))` using the `&` operator was confusing given that `and` is used as well.

File: infra/cifuzz/cifuzz.py [hunk 10]
```diff
@@ -320,24 +323,70 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
+      affected_fuzzers.append(os.path.basename(fuzzer))
+
+  if not affected_fuzzers:
+    logging.info('Not using affected fuzzers.')
+    return
+  logging.info(
+      'Using affected fuzzers.\n The following fuzzers are affected: %s',
+      ' '.join(affected_fuzzers))
+
+  # Remove all the fuzzers that are not affected
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz.py]: period.

File: infra/cifuzz/cifuzz.py [hunk 11]
```diff
@@ -320,24 +323,70 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
+                                                src_in_docker)
+
+    if covered_files and set(covered_files) & set(files_changed):
+      affected_fuzzers.append(os.path.basename(fuzzer))
+
+  if not affected_fuzzers:
+    logging.info('Not using affected fuzzers.')
+    return
+  logging.info(
+      'Using affected fuzzers.\n The following fuzzers are affected: %s',
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz.py]: Why a space after the newline?
It might be cleaner to write this on two lines
```
`'Using affected fuzzers.\n'
'The following fuzzers are affected: %s',
```

Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: done.

File: infra/cifuzz/cifuzz.py (hunk 10 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Done.

File: infra/cifuzz/cifuzz.py (hunk 1 above)
Comment (Dor1s) [reply infra/cifuzz/cifuzz.py]: nit: what is "project base"? can the message be more descriptive?

and for the coverage part, let's also be more explicit, e.g. "Can't get coverage information from OSS-Fuzz."

File: infra/cifuzz/cifuzz.py [hunk 12]
```diff
@@ -319,25 +322,75 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
 
 
+def remove_unaffected_fuzzers(project_name, out_dir, files_changed,
+                              src_in_docker):
+  """Removes all non affected fuzzers in the out directory.
+
+  Args:
+    project_name: The name of the relevant OSS-Fuzz project.
+    out_dir: The location of the fuzzer binaries.
+    files_changed: A list of files changed from origin.
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: nit: I'm not sure "origin" is the good word here, as it has a certain meaning in git. Let's better say either "changed in a pull request." or "changed compared to HEAD."

File: infra/cifuzz/cifuzz.py [hunk 13]
```diff
@@ -319,25 +322,75 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
+  if not latest_cov_report_info:
+    logging.error('Could not download latest coverage report.')
+    return
+  affected_fuzzers = []
+  for fuzzer in fuzzer_paths:
+    covered_files = get_files_covered_by_target(latest_cov_report_info,
+                                                os.path.basename(fuzzer),
+                                                src_in_docker)
+
+    if covered_files and set(covered_files).intersection(set(files_changed)):
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: I don't think we should skip fuzzers which we don't have coverage for. I'd write this as follows:

```python
if not covered files:
  # Assume a fuzzer is affected if we can't get its coverage from OSS-Fuzz.
  affected_fuzzers.append(os.path.basename(fuzzer))
  continue

covered_files = set(covered_files)  # Are you sure there might be duplicates, btw? I doubt it.
for file in files_changed:
  if file in covered_files:
      affected_fuzzers.append(os.path.basename(fuzzer))
      continue
```

File: infra/cifuzz/cifuzz.py [hunk 14]
```diff
@@ -319,25 +322,75 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
+
+  if not affected_fuzzers:
+    logging.info('No affected fuzzers detected, keeping all as fallback.')
+    return
+  logging.info(
+      'Using affected fuzzers.\n The following fuzzers are affected: %s',
+      ' '.join(affected_fuzzers))
+
+  # Remove all the fuzzers that are not affected.
+  for file in os.listdir(out_dir):
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: looks like this would also remove other files, which might include important shared libraries (some projects use dynamic linking despite our advise not to). Please re-write this to match the comment and delete only non affected fuzzer binaries

File: infra/cifuzz/cifuzz_test.py [hunk 15]
```diff
@@ -361,5 +361,54 @@ def test_get_invalid_project(self):
     self.assertIsNone(cifuzz.get_latest_cov_report_info(''))
 
 
+class KeepAffectedFuzzersUnitTest(unittest.TestCase):
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz_test.py]: also add a test for a case where we don't have coverage for a fuzzer. In real like it could be a new fuzz target which was added or renamed just today

File: infra/cifuzz/cifuzz.py (hunk 1 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Fixed.

File: infra/cifuzz/cifuzz.py (hunk 12 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Done.

File: infra/cifuzz/cifuzz.py (hunk 13 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: I agree about keeping fuzzers without coverage reports. I was wondering why you chose the iteration comparison rather than set comparison though?

File: infra/cifuzz/cifuzz.py (hunk 14 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Added.

File: infra/cifuzz/cifuzz_test.py (hunk 15 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz_test.py]: Added.

File: infra/cifuzz/cifuzz.py [hunk 16]
```diff
@@ -319,25 +323,79 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
 
-    relative_path = file['filename'].replace(oss_fuzz_project_base, '')
+    relative_path = file['filename'].replace(norm_src_in_docker, '')
     affected_file_list.append(relative_path)
   if not affected_file_list:
     return None
   return affected_file_list
 
 
+def remove_unaffected_fuzzers(project_name, out_dir, files_changed,
```
Comment (jonathanmetzman) [inline infra/cifuzz/cifuzz.py]: I would just keep a list of affected fuzzers and run those rather than deleting each unaffected fuzzer.

File: infra/cifuzz/cifuzz.py (hunk 13 above)
Comment (Dor1s) [reply infra/cifuzz/cifuzz.py]: using set comparison is redundant, because
a) creating a set isn't necessary
b) intersection iterates through all elements, while it's sufficient to find just one common element

File: infra/cifuzz/cifuzz.py (hunk 12 above)
Comment (Dor1s) [reply infra/cifuzz/cifuzz.py]: doesn't seem to be addressed, same for line 358, I don't think `origin` is a good word there too

File: infra/cifuzz/cifuzz.py [hunk 17]
```diff
@@ -319,25 +323,79 @@ def get_files_covered_by_target(latest_cov_info, target_name,
...
+  for fuzzer in fuzzer_paths:
+    covered_files = get_files_covered_by_target(latest_cov_report_info,
+                                                os.path.basename(fuzzer),
+                                                src_in_docker)
+    if not covered_files:
+      # Assume a fuzzer is affected if we can't get its coverage from OSS-Fuzz.
+      affected_fuzzers.append(os.path.basename(fuzzer))
+      continue
+
+    if covered_files and set(covered_files).intersection(set(files_changed)):
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: `if covered_files` check is redundant, as we have `if not covered_files` followed by `continue` just before this, i.e. `if covered_files` is always `True` here

File: infra/cifuzz/cifuzz.py (hunk 13 above)
Comment (Dor1s) [reply infra/cifuzz/cifuzz.py]: hm I think I've just posted a reply to this but I don't see it, let me repeat just in case: set comparison is redundant because:
a) we don't really need to create sets here
b) intersection iterates through all elements, while we need to find at least one common across two lists

File: infra/cifuzz/cifuzz.py (hunk 5 above)
Comment (Dor1s) [reply infra/cifuzz/cifuzz.py]: +1, the message isn't very descriptive

File: infra/cifuzz/cifuzz.py (hunk 11 above)
Comment (Dor1s) [reply infra/cifuzz/cifuzz.py]: same here, better to say "%d fuzzer seem to be affected by the pull request."

File: infra/cifuzz/cifuzz.py (hunk 14 above)
Comment (Dor1s) [reply infra/cifuzz/cifuzz.py]: Why do we need `os.listdir()` if we know names of files we want to delete?

File: infra/cifuzz/cifuzz.py (hunk 16 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: That makes the run_fuzzers action interlocked with the build_fuzzers action though. I think it would be good to make the run fuzzers action as dumb as possible.

File: infra/cifuzz/cifuzz.py (hunk 11 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: I think having the names will be helpful for debugging.

File: infra/cifuzz/cifuzz.py (hunk 12 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Fixed.

File: infra/cifuzz/cifuzz.py (hunk 13 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Makes sense. Fixed.

File: infra/cifuzz/cifuzz.py (hunk 14 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Fixed.

File: infra/cifuzz/cifuzz.py [hunk 18]
```diff
@@ -163,6 +163,10 @@ def build_fuzzers(project_name,
   if helper.docker_run(command):
     logging.error('Building fuzzers failed.')
     return False
+
+  # Remove non-affected fuzzers.
```
[the remark this answers is not in the document]
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Removed.

File: infra/cifuzz/cifuzz.py (hunk 17 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Fixed.

File: infra/cifuzz/cifuzz.py (hunk 5 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Changed to Oliver's suggestion.

File: infra/cifuzz/cifuzz.py (hunk 16 above)
Comment (jonathanmetzman) [reply infra/cifuzz/cifuzz.py]: Maybe, this still feels kind of hacky but i'm not sure what's better.
